绘制相关代码

Moon-404 opened the pull request:
代码可以运行了，但是表现上还有点问题，我是没有更好的实现方式了，需要更多技术支持

Moon-404 committed:
绘制UI事件（存在bug）

运行时会全部卡死，需要debug

Moon-404 committed:
绘制文字相关代码

可以运行，但是表现上还存在问题，需要进一步开发

Violet-Nonbloosom commented:
测试出来所谓“漏洞”的原因了。绘制代码没有问题，但是绘制事件的状况欠佳。
绘制事件高频触发，使得 PVZ 本体无法运行。
而如果尝试用 Sleep() 进行延时，虽然 PVZ 本体没事了，但绘制过程前功尽弃，这些代码就废了。
总之，我不同意合并这些代码，你可以接着改代码了。

Moon-404 committed:
一种新形式的贴字

还有效率问题需要解决

Moon-404 commented:
> 测试出来所谓“漏洞”的原因了。绘制代码没有问题，但是绘制事件的状况欠佳。 绘制事件高频触发，使得 PVZ 本体无法运行。 而如果尝试用 Sleep() 进行延时，虽然 PVZ 本体没事了，但绘制过程前功尽弃，这些代码就废了。 总之，我不同意合并这些代码，你可以接着改代码了。

我现在的思路是，在每一次刷新画面前中断，由pvzclass把所有需要输出的字保存到结构体中，由pvz主程序依次处理。
我最新的commit有这个实现方式的例子，但它还远没有完成，暂时不要合并，但我觉得你可以看一下效果。
使用方式是，将贴字.CT中的贴字(pvzclass)启用，然后启动我这个commit里面的pvzclass文件，效果是在关卡中显示植物的生命值。
问题在于58EB50这段被注释掉的代码会导致PVZ卡顿，但是不加上它又会使得无法调整字体大小，需要一些新的灵感。
在找到合适的解决方案后，我会移除无效的代码，并使其可以仅通过pvzclass实现。

![image](https://github.com/Lazuplis-Mei/pvzclass/assets/33384538/e129efb1-68ee-4467-89f7-24d3aca8ba7c)

Violet-Nonbloosom commented:
我目前的测试情况也是这样。
我的话，可能会考虑写一个汇编脚本生成器吧。

Moon-404 commented:
如果使用 6A72EC 这个 ImageFont，则可以默认不提供修改字体大小的功能，因为它的初始大小还是比较合适的。
另外，经测试，原来是 50FPS 的场景，在使用绘制字符串时帧数会大致减半，我不确定这是否足够高效。
还需要更多 ImageFont 常量地址的资料，以便测试出最适合用来绘制字的字体。另外，也许可以优化一下汇编码以提高效率。

Moon-404 committed:
Update 贴字.CT

更新一种贴字方式，字体比较舒服，移除占用较大资源的调节字体大小功能。
制作时会给用户提供初始化接口，一切以提高效率为主。
在50FPS的游戏场景下，这段示例代码会降低一半的帧率，需要更高效的实现方式？

Moon-404 committed:
绘制字符串更新

在开启这段代码进行游戏时，砸罐子关卡可以达到30FPS，帧数与需要绘制的字符串数量成负相关关系。
测试发现两个问题：
1. 植物数量发生变化时字体也会变化，程序的修改被屏蔽了；
2. 关卡通关后再打开新的关卡，原来的数据不会被清除，但是理论上如果num为0确实会清空数据的。
其它修改：
1. 增加一些ASM宏
2. 移除Graphics
3. AllocMemory增加page数量的参数，默认为1
4. CreateThread不再改变processId，这个改变是无意义的
5. 更新了Draw和DrawUITopEvent

Moon-404 commented:
现在这个基本上就是终版了，手头上有点别的事，测试出的问题等我空下来再看看能不能解决。
可以先看一看，等我把问题解决再合并，万一有我没测出来的问题也说不定呢。

Moon-404 committed:
PlantReloadEvent

Moon-404 committed:
修复渲染过时字符串的bug

Moon-404 committed:
增加子弹命中僵尸事件

Violet-Nonbloosom committed:
Merge branch 'master' into master
